middleware/http_cluster.go

```diff
@@ -0,0 +1,148 @@
+package middleware
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/dskit/clusterutil"
+	"github.com/grafana/dskit/httpgrpc"
+)
+
+type clusterValidationError struct {
+	ClusterValidationErrorMessage string `json:"cluster_validation_error_message"`
+}
+
+// writeAsJSON writes this error as JSON to the HTTP response.
+func (e *clusterValidationError) writeAsJSON(w http.ResponseWriter) {
+	data, err := json.Marshal(e)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusNetworkAuthenticationRequired)
+	// We ignore errors here, because we cannot do anything about them.
+	// Write will trigger sending Status code, so we cannot send a different status code afterwards.
+	// Also, this isn't internal error, but error communicating with client.
+	_, _ = w.Write(data)
+}
+
+func ClusterValidationRoundTripper(cluster string, invalidClusterValidationReporter InvalidClusterValidationReporter, next http.RoundTripper) RoundTripperFunc {
+	validateClusterValidationRoundTripperInputParameters(cluster, invalidClusterValidationReporter)
+	return func(req *http.Request) (*http.Response, error) {
+		clusterutil.PutClusterIntoHeader(req, cluster)
+		resp, err := next.RoundTrip(req)
+		if err != nil {
+			return nil, err
+		}
+		if resp.StatusCode != http.StatusNetworkAuthenticationRequired {
+			return resp, nil
+		}
+		if resp.Header.Get("Content-Type") != "application/json" {
+			return resp, nil
+		}
+		defer resp.Body.Close()
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, err
+		}
+		var clusterValidationErr clusterValidationError
+		err = json.Unmarshal(body, &clusterValidationErr)
```

Comment:
Come to think of it, we probably want to protect ourselves from a too long response body:
```suggestion
		var clusterValidationErr clusterValidationError
		err = json.NewDecoder(io.LimitReader(resp.Body, 1024)).Decode(&clusterValidationErr)
```

Reply:
As agreed, this will be done in a follow-up.